doc/reference/trd104-syscalls.md

```diff
@@ -303,61 +303,108 @@ call to invoke. The kernel invokes upcalls only in response to Yield
-When the `yield-wait` returns, the process checks if the resuming upcall was
-the one it was expecting, and if not calls `yield-wait` again.
-
-The second call, `yield-no-wait`, executes a single upcall if any is pending.
-If no upcalls are pending it returns immediately.
+There are three Yield system calls:
+  - Yield-Wait
+  - Yield-NoWait
+  - Yield-WaitFor
 
 The register arguments for Yield system calls are as follows. The registers
 r0-r3 correspond to r0-r3 on CortexM and a0-a3 on RISC-V.
 
 | Argument               | Register |
 |------------------------|----------|
 | Yield number           | r0       |
-| No wait field          | r1       |
-| unused                 | r2       |
-| unused                 | r3       |
+| yield-param-1          | r1       |
+| yield-param-2          | r2       |
+| yield-param-3          | r3       |
 
+The Yield system call class has no return value. This is because
+invoking an upcall pushes that function call onto the stack, such
+that the return value of a call to yield system call may be the
+return value of the upcall.
 
-The yield number specifies which call is invoked.
+Yield variants which return a result to the caller will use one or more
+of the `yield-param-N` arguments to do so.  Allowing the kernel to pass
+a return value in a register back to userspace would require either
+re-entering the kernel or expensive execution architectures (e.g.,
+additional stacks or additional stack frames) for upcalls.
+
+The Yield number (in r0) specifies which call is invoked:
 
 | System call     | Yield number value |
 |-----------------|--------------------|
 | yield-no-wait   |                  0 |
 | yield-wait      |                  1 |
-
+| yield-wait-for  |                  2 |
 
 All other yield number values are reserved. If an invalid
 yield number is passed the kernel MUST return immediately.
 
-The no wait field is only used by `yield-no-wait`. It contains the
-memory address of an 8-bit byte that `yield-no-wait` writes to
-indicate whether an upcall was invoked. If invoking `yield-no-wait`
-resulted in an upcall executing, `yield-no-wait` writes 1 to the
-field address. If invoking `yield-no-wait` resulted in no upcall
-executing, `yield-no-wait` writes 0 to the field address. This field
-allows userspace loops that want to flush the upcall queue to
-execute `yield-no-wait` until the queue is empty.
+The meaning of `yield-param-N` (r1-r3) is specific to the yield type.
+
+
+### 4.1.1 Yield-NoWait
+
+Yield number 0, Yield-NoWait, executes a single upcall if any is
+pending.  If no upcalls are pending it returns immediately.
+
+`yield-param-1` contains the memory address of an 8-bit byte that
+Yield-NoWait writes to indicate whether an upcall was invoked. If
+invoking Yield-NoWait resulted in an upcall executing, Yield-NoWait
+writes 1 to the field address. If invoking Yield-NoWait resulted in no
+upcall executing, Yield-NoWait writes 0 to the field address. This field
```

Comment:
I don't understand why we use pointers for yield-nowait and return values for yield-waitfor. Is there an optimization for the yield-nowait case that made the pointer the correct choice?

Reply:
When the kernel returns to userspace, it only gets to set registers r0-r3 for one stack frame. In practice, we have two cases:
 - "DirectResume" {command, subscribe, allow, memop; yield-no-wait_DidntWaitCase}: We are resuming execution directly after the `svc` invocation, so the assembly that follows the `svc` command can use the values in r0-r3 as-set by the kernel
 - "PushedCallback" {yield-wait, yield-no-wait_DidWaitCase}: We resume execution at the start of the callback function. The values in r0-r3 are _consumed_ by the callback. When the callback finishes, it will `pop {lr}`, which has been set by the kernel to the instruction after the `svc`, and the assembly that invoked the syscall now gets to run. At this point `r0-r3` are _unknown_ as those are caller-save registers, so the callback can clobber them freely. The assembly that invoked the `svc` cannot make any assumptions about the values in `r0-r3`, nor can the kernel use them to pass things "to" the calling assembly. Thus, the "PushedCallback" case has to use the pointer approach for the kernel to communicate with the assembly that invokes the `svc`.

For Yield-WaitFor as-written, we again have two cases.
 - There _was not_ a subscribed callback, so we "DirectResume" and the kernel 'passes arguments to the calling assembly'
 - There _was_ a subscribed callback, so we are in the "PushedCallback" case, and the assembly that invoked the YieldFor can't get any information from the kernel.

Reply:
Thanks that is very clear and it would be great to capture it somewhere.